compiler/src/dotty/tools/dotc/semanticdb/TypeOps.scala

```diff
@@ -0,0 +1,305 @@
+            case RefinedType(parent, name, info) =>
+              flatten(parent, acc :+ (name, info))
+            case _ =>
+              (tpe, acc)
+          }
+
+          // flatten parent types to list
+          // e.g. `X with Y with Z { refined }`
+          // RefinedType(parent = AndType(X, AndType(Y, Z)), ...)
+          // => List(X, Y, Z)
+          def flattenParent(parent: Type): List[s.Type] = parent match {
+            case AndType(tp1, tp2) =>
+              flattenParent(tp1) ++ flattenParent(tp2)
+            case _ => List(loop(parent))
+          }
+
+          val (parent, refinedInfos) = flatten(rt, List.empty)
+          val stpe = s.WithType(flattenParent(parent))
+
+          val decls = refinedInfos.flatMap { (name, _) =>
+            refinementSymtab.get((rt, name))
+          }
+          val sdecls = decls.sscope(using LinkMode.HardlinkChildren)
+          s.StructuralType(stpe, Some(sdecls))
+
+        case rec: RecType =>
+          loop(rec.parent) // should be handled as RefinedType
+
+        // repeated params: e.g. `Int*`, which is the syntax sugar of
+        // `Seq[Int] @Repeated` (or `Array[Int] @Repeated`)
+        // See: Desugar.scala and TypeApplications.scala
+        case AnnotatedType(AppliedType(_, targs), annot)
+          if (annot matches defn.RepeatedAnnot) && (targs.length == 1) =>
+          val stpe = loop(targs(0))
+          s.RepeatedType(stpe)
+
+        case ann: AnnotatedType if ann.annot.symbol.info.isInstanceOf[ClassInfo] =>
+          def flatten(tpe: Type, annots: List[Annotation]): (Type, List[Annotation]) = tpe match
+            case AnnotatedType(parent, annot) if annot.symbol.info.isInstanceOf[ClassInfo] =>
+              flatten(parent, annot +: annots)
+            case other => (other, annots)
+
+          val (parent, annots) = flatten(ann, List.empty)
+          val sparent = loop(parent)
+          val sannots = annots.map(a =>
+            s.Annotation(loop(a.symbol.info.asInstanceOf[ClassInfo].selfType))
+          )
+          s.AnnotatedType(sannots, sparent)
+
+        case app @ AppliedType(tycon, args) =>
+          val sargs = args.map(loop)
+          loop(tycon) match
+            case ref: s.TypeRef => ref.copy(typeArguments = sargs)
+            // is there any other cases?
+            case _ => s.Type.Empty
+
+        case and: AndType =>
+          def flatten(child: Type): List[Type] = child match
+            case AndType(ct1, ct2) => flatten(ct1) ++ flatten(ct2)
+            case other => List(other)
+          val stpes = flatten(and).map(loop)
+          s.IntersectionType(stpes)
+
+        case or: OrType =>
+          def flatten(child: Type): List[Type] = child match
+            case OrType(ct1, ct2) => flatten(ct1) ++ flatten(ct2)
+            case other => List(other)
+          val stpes = flatten(or).map(loop)
+          s.UnionType(stpes)
+
+        case NoPrefix =>
+          s.Type.Empty
+
+        case _ => s.Type.Empty
```

Comment:
Addressed by https://github.com/lampepfl/dotty/pull/12885/commits/f509613cfd5ece7bd31789c57c907d2388040763
Made a pattern match for `TypeBounds` inside the `AppliedType` because we have to convert `AppliedType` to `ExistentialType` in some cases, and I assume `TypeBounds` appears only in `AplliedType` afaik (and I confirmed at least in the expected tests, there are no `TypeBounds` outside of `AppliedType`).